workers/unity/Packages/io.improbable.gdk.tools/PortForwardingWindow.cs

```diff
@@ -0,0 +1,187 @@
+            using (new EditorGUILayout.VerticalScope())
+            {
+                deploymentName = EditorGUILayout.TextField("Deployment Name", deploymentName);
+                workerId = EditorGUILayout.TextField("Worker ID", workerId);
+                port = EditorGUILayout.IntField("Port", port);
+
+                var errors = GetErrors().ToList();
+
+                if (errors.Any())
+                {
+                    EditorGUILayout.HelpBox(
+                        $"Cannot forward port due to the following errors:\n\n{string.Join("\n", errors)}",
+                        MessageType.Error);
+                }
+
+                using (new EditorGUI.DisabledScope(errors.Any()))
+                {
+                    if (GUILayout.Button("Forward port"))
+                    {
+                        portForwardingHasFailed = false;
+                        isConnected = false;
+                        activePortForwarding = RunPortForwarding();
+                    }
+                }
+            }
+
+            var messageType = MessageType.None;
+            string message = null;
+
+            if (isConnected)
+            {
+                messageType = MessageType.Info;
+                message = $"Connected successfully to {workerId}:{port} on {deploymentName}!\n\nAssembly reloading locked.";
+            }
+            else if (activePortForwarding != null)
+            {
+                messageType = MessageType.Warning;
+                message = "Establishing connection...\n\nAssembly reloading locked.";
+            }
+            else if (portForwardingHasFailed)
+            {
+                messageType = MessageType.Error;
+                message = "Failed to establish connection. Check the Unity Console for more information.";
+            }
+
+            if (message != null)
+            {
+                EditorGUILayout.HelpBox(message, messageType);
+            }
+
+            if (activePortForwarding != null)
+            {
+                if (GUILayout.Button("Cancel"))
+                {
+                    tokenSrc.Cancel();
+                }
+            }
+        }
+
+        private void OnDisable()
+        {
+            Cleanup();
+        }
+
+        private void OnDestroy()
+        {
+            Cleanup();
+        }
+
+        private void Update()
+        {
+            if (activePortForwarding == null)
+            {
+                return;
+            }
+
+            if (!activePortForwarding.IsCompleted)
+            {
+                return;
+            }
+
+            // Cancelling or killing the process will result in a non-zero exit code, so ensure that we weren't
+            // previously connected successfully.
+            portForwardingHasFailed = activePortForwarding.Result.ExitCode != 0 && !isConnected;
+
+            if (portForwardingHasFailed)
+            {
+                foreach (var error in activePortForwarding.Result.Stdout)
```

Comment:
Do we not want to concatenate this into 1 error?

Reply:
Discussed offline - increases noise from the error messaging